File: app/src/main/java/com/widiarifki/screentest/presentation/guest/GuestPresenter.java
package com.widiarifki.screentest.presentation.guest;

import com.widiarifki.screentest.model.Guest;
import com.widiarifki.screentest.services.GuestService;
import com.widiarifki.screentest.services.ServiceBuilder;

import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

/**
 * Created by widiarifki on 27/02/2018.
 */

public class GuestPresenter {
    IGuestView mView;

    public GuestPresenter(IGuestView view) {
        mView = view;
    }

    public void fetchGuestData(){
        mView.showProgressDialogue(true);

        GuestService guestService = ServiceBuilder.buildService(GuestService.class);
        Call<List<Guest>> guestRequest = guestService.getGuest();
        guestRequest.enqueue(new Callback<List<Guest>>() {
            @Override
            public void onResponse(Call<List<Guest>> request, Response<List<Guest>> response) {
                mView.showList(response.body());
            }

            @Override
            public void onFailure(Call<List<Guest>> request, Throwable t) {
                mView.showGetDataFail();
            }
        });

        /*OkHttpClient httpClient = new OkHttpClient();
        Request httpReq = new Request.Builder().url("http://dry-sierra-6832.herokuapp.com/api/people").build();
        Call httpCall = httpClient.newCall(httpReq);
        httpCall.enqueue(new Callback() {
            @Override
            public void onFailure(Call call, IOException e) {
                mView.showGetDataFail();
            }

            @Override
            public void onResponse(Call call, Response response) throws IOException {
                if(response.isSuccessful() && response.code() == 200){
                    String dataString = response.body().string();
                    final List<Guest> guestList = new ArrayList<>();
                    try {
                        JSONArray dataJson = new JSONArray(dataString);
                        if(dataJson.length() > 0){
                            for(int i = 0; i < dataJson.length(); i++){
                                JSONObject record = dataJson.getJSONObject(i);
                                Guest guest = new Guest();
                                guest.setId(record.getInt("id"));
                                guest.setName(record.getString("name"));
                                guest.setBirthdate(record.getString("birthdate"));
                                guestList.add(guest);
                            }

                            mView.showList(guestList);
                        }
                    } catch (JSONException e) {
                        e.printStackTrace();
                    }

                }
            }
        });*/
    }
}

File: app/src/main/java/com/widiarifki/screentest/presentation/guest/GuestPresenter.kt
package com.widiarifki.screentest.presentation.guest

import com.widiarifki.screentest.model.Guest
import com.widiarifki.screentest.services.GuestService
import com.widiarifki.screentest.services.ServiceBuilder
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import io.realm.Realm



/**
 * Created by widiarifki on 27/02/2018.
 */

class GuestPresenter(internal var mView: IGuestView) {

    fun fetchGuestData() {

        val apiService = ServiceBuilder.buildService(GuestService::class.java)
        val realm = Realm.getDefaultInstance()
        val realmQuery = realm.where(Guest::class.java)
        val savedList: List<Guest>? = realmQuery.findAll()

        //get user if it's already saved
        if (savedList != null) {
            mView.showList(savedList)
        }

        apiService.getGuests()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(
                        { guestList ->
                            realm.beginTransaction()
                            realm.copyToRealmOrUpdate(guestList)
                            realm.commitTransaction()
                            mView.showList(guestList)
                        },
                        { error ->
                            if(savedList != null && savedList.count() == 0) {
                                mView.showDialogMessage("Something's Wrong", "Fetching data from server failed")
                            }
                        }
                )
    }

}
